sentry/src/main/java/io/sentry/SentryOptions.java

```diff
@@ -1,60 +1,311 @@
+import io.sentry.util.Nullable;
+import io.sentry.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * SentryOptions used to configure the Sentry SDK.
  */
-public class SentryOptions {
-    private SentryClientFactory sentryClientFactory;
-    private String dsn;
-    private ResourceLoader resourceLoader;
+public final class SentryOptions {
+    private static final Logger logger = LoggerFactory.getLogger(SentryOptions.class);
+
+    private final Lookup lookup;
+    private final Dsn dsn;
+    private final SentryClientFactory clientFactory;
+
+    /**
+     * Creates a new instance of the options using the provided parameters.
+     *
+     * @param lookup        the lookup to locate the configuration with
+     * @param dsn           the dsn to use
+     * @param clientFactory the client factory to use
+     * @throws NullPointerException if any of the parameters is null
+     */
+    public SentryOptions(Lookup lookup, Dsn dsn, SentryClientFactory clientFactory) {
+        this.lookup = requireNonNull(lookup, "lookup");
+        this.dsn = requireNonNull(dsn, "dsn");
+        this.clientFactory = requireNonNull(clientFactory, "clientFactory");
+    }
+
+    /**
+     * Creates new options using the provided lookup instance. The DSN and the client factory are obtained using the
+     * provided lookup instance.
+     *
+     * @param lookup the lookup to locate the configuration with
+     * @return new instance of SentryOptions
+     */
+    public static SentryOptions from(Lookup lookup) {
+        return from(lookup, (String) null, null);
+    }
+
+    /**
+     * Creates new options using the provided lookup instance and optional DSN. If the provided DSN is null, the DSN
+     * to use is looked up using the provided lookup instance.
+     *
+     * @param lookup the lookup to locate the configuration with
+     * @param dsn    the optional dsn to use
+     * @return new instance of SentryOptions
+     */
+    public static SentryOptions from(Lookup lookup, @Nullable String dsn) {
+        return from(lookup, dsn, null);
+    }
+
+    /**
+     * Creates new options using the provided lookup instance and optional DSN and client factory. If the provided
+     * DSN or client factory is null, the DSN and client factory to use is looked up using the provided lookup instance.
+     *
+     * @param lookup  the lookup to locate the configuration with
+     * @param dsn     the optional dsn to use or null if the value should be located in the config
+     * @param factory the client factory to use or null if the value should be located in the config
+     * @return new instance of SentryOptions
+     */
+    public static SentryOptions from(Lookup lookup, @Nullable String dsn, @Nullable SentryClientFactory factory) {
+        return from(lookup, resolveDsn(lookup, dsn), factory);
+    }
+
+    private static SentryOptions from(Lookup lookup, Dsn dsn, @Nullable SentryClientFactory factory) {
```

Comment:
By not exposing `Dsn` we get rid of some of these overloads too. Reduce the API surface.